src/Nethermind/Nethermind.Baseline/JsonRpc/BaselineModule.cs

```diff
@@ -0,0 +1,307 @@
+                new BlockParameter(0L),
+                new BlockParameter(_blockFinder.Head.Number),
+                new AddressFilter(treeAddress),
+                new TopicsFilter(new SpecificTopic(leafTopic))); // find tree topics
+
+            var insertLeavesLogs = _logFinder.FindLogs(insertLeavesFilter);
+            var insertLeafLogs = _logFinder.FindLogs(insertLeafFilter);
+            BaselineTree baselineTree = new ShaBaselineTree(new MemDb(), new byte[0], 5);
+
+            // Keccak leafTopic = new Keccak("0x8ec50f97970775682a68d3c6f9caedf60fd82448ea40706b8b65d6c03648b922");
+            foreach (FilterLog filterLog in insertLeavesLogs)
+            {
+                for (int i = 0; i < (filterLog.Data.Length - 128) / 32; i++)
+                {
+                    Bytes32 leafHash = Bytes32.Wrap(filterLog.Data.Slice(128 + 32 * i, 32).ToArray());
+                    baselineTree.Insert(leafHash);
+                }
+            }
+
+            foreach (FilterLog filterLog in insertLeafLogs)
+            {
+                Bytes32 leafHash = Bytes32.Wrap(filterLog.Data.Slice(32, 32).ToArray());
+                baselineTree.Insert(leafHash);
+            }
+
+            return baselineTree;
+        }
+
+        public Task<ResultWrapper<BaselineTreeNode[]>> baseline_getSiblings(Address contractAddress, long leafIndex)
+        {
+            if (leafIndex > MerkleTree.MaxLeafIndex || leafIndex < 0L)
+            {
+                return Task.FromResult(ResultWrapper<BaselineTreeNode[]>.Fail($"{leafIndex} is not a valid leaf index", ErrorCodes.InvalidInput));
+            }
+
+            bool isTracked = _baselineTrees.TryGetValue(contractAddress, out BaselineTree? tree);
+            if (!isTracked)
+            {
+                return Task.FromResult(ResultWrapper<BaselineTreeNode[]>.Fail($"{contractAddress} tree is not tracked", ErrorCodes.InvalidInput));
+            }
+
+            // everything in memory
+            tree = RebuildEntireTree(contractAddress);
+
+            return Task.FromResult(ResultWrapper<BaselineTreeNode[]>.Success(tree!.GetProof((uint) leafIndex)));
+        }
+
+        public Task<ResultWrapper<bool>> baseline_track(Address contractAddress)
+        {
+            // can potentially warn user if tree is not deployed at the address
+
+            if (TryAddTree(contractAddress))
+            {
+                UpdateMetadata(contractAddress);
+                return Task.FromResult(ResultWrapper<bool>.Success(true));
+            }
+
+            return Task.FromResult(ResultWrapper<bool>.Fail($"{contractAddress} is already tracked", ErrorCodes.InvalidInput));
+        }
+
+        public Task<ResultWrapper<Address[]>> baseline_getTracked()
+        {
+            return Task.FromResult(ResultWrapper<Address[]>.Success(_metadata.TrackedTrees));
+        }
+
+        private void UpdateMetadata(Address contractAddress)
+        {
+            var list = _metadata.TrackedTrees.ToList();
```

Comment:
Why ToList/ToArray?

Reply:
not thread safe anyway so needs a change / lock
ToList ToArray - unipmportant, can do Array.Resize - not that much different as the method shall be called very rarely and the list will not be long